### ISSUE TITLE
Fix "residual error" mentioned on line 213 of OpenALRenderer.cpp

### ISSUE THREAD

**Issue by stephengtuggy** [vegastrike/Vega-Strike-Engine-Source#859]
So. An exception was thrown on line 215 of OpenALRenderer.cpp. That's a bummer.

Line 213 of that file (2 lines further up) has a comment that says the following:

```c++14
// FIXME: There's a residual error here on Windows. Can't track down where it's from.
```

This is not encouraging LOL. I guess the `residual error` mentioned also occurs on macOS X 10.13.

I would think we should treat this crash as a separate issue. I'll go ahead and file one.

@daviewales can you try setting Sounds to No Sound, and audio volume to Off, in vegasettings? Then save your changes, and try running the game again? See if that makes any difference.

_Originally posted by @stephengtuggy in https://github.com/vegastrike/Vega-Strike-Engine-Source/issues/533#issuecomment-2094987065_

**Comment by BenjamenMeyer** [comment]
do we consider this a 0.9.x or 0.10.x issue?

**Comment by daviewales** [comment]
I tried disabling sound in `vegasettings`, but I still get the same OpenAL error.

**Comment by daviewales** [comment]
The binary is linked correctly to the MacPorts OpenAL library:

```
otool -L vegastrike-engine.app/Contents/MacOS/vegastrike-engine | grep -i openal
	/opt/local/lib/libopenal.1.dylib (compatibility version 1.0.0, current version 1.23.1)
```

**Comment by stephengtuggy** [comment]
> do we consider this a 0.9.x or 0.10.x issue?

@BenjamenMeyer I'm thinking 0.10.x. Or 0.9.1, at a minimum. Who knows how long this problem has been there, or how long it will take to fix.

**Comment by stephengtuggy** [comment]
@daviewales Another thing you can try is to find the line in engine/CMakeLists.txt that sets `HAVE_AL` to 1, and modify that line to set it to 0 instead. You should then (hopefully) get a build of the game compiled without audio support from the get-go.

Sorry that you're running into this issue. Hopefully it doesn't turn out to be too difficult to fix.

**Comment by daviewales** [comment]
I set `HAVE_AL` to 0, but I'm still getting:

```
OpenAL error: unknown(Bad value) at /Users/dwales/Code/not-my-code/Vega-Strike-Engine-Source/engine/src/audio/renderers/OpenAL/OpenALRenderer.cpp:215
libc++abi.dylib: terminating with uncaught exception of type Audio::Exception: OpenAL error: unknown(Bad value) at /Users/dwales/Code/not-my-code/Vega-Strike-Engine-Source/engine/src/audio/renderers/OpenAL/OpenALRenderer.cpp:215
```

HOWEVER, the following diff prevents the error and plays without crashing!!
(Not sure why it works, or if it's a good idea, but I can fire lasers from a spaceship now!)

``` diff
diff --git a/engine/src/audio/renderers/OpenAL/OpenALRenderer.cpp b/engine/src/audio/renderers/OpenAL/OpenALRenderer.cpp
index 606882fb7..aa89e36bd 100644
--- a/engine/src/audio/renderers/OpenAL/OpenALRenderer.cpp
+++ b/engine/src/audio/renderers/OpenAL/OpenALRenderer.cpp
@@ -162,7 +162,7 @@ struct RendererData {
         alDevice = alcOpenDevice((ALCstring) (deviceSpecifier));
 
         if (!alDevice)
-            checkAlError();
+            {/* checkAlError(); */}
         else
             clearAlError();
     }
@@ -196,26 +196,26 @@ struct RendererData {
 
         alContext = alcCreateContext(alDevice, params);
         if (!alContext)
-            checkAlError();
+            {/* checkAlError(); */}
         else
             clearAlError();
 
         alcMakeContextCurrent(alContext);
-        checkAlError();
+        /* checkAlError(); */
     }
 
     void commit() {
         alcProcessContext(alContext);
-        checkAlError();
+        /* checkAlError(); */
     }
 
     void suspend() {
         // FIXME: There's a residual error here on Windows. Can't track down where it's from.
         alGetError();
-        checkAlError();
+        /* checkAlError(); */
         alcMakeContextCurrent(alContext);
         alcSuspendContext(alContext);
-        checkAlError();
+        /* checkAlError(); */
     }
 
     void closeContext() {
@@ -381,7 +381,7 @@ void OpenALRenderer::setupDopplerEffect() {
 
     data->dirty.dopplerFactor = 0;
 
-    checkAlError();
+    /* checkAlError(); */
 }
 
 BorrowedOpenALRenderer::BorrowedOpenALRenderer(ALCdevice *device, ALCcontext *context) :
```

**Comment by stephengtuggy** [comment]
@daviewales Hmm. Very interesting to know that the above diff allows so many things to start working. I would hesitate to apply the diff just yet, though, until we know more. It could be that commenting out the error checks simply masks some errors that are important to know about. (And should be solved some other way.)

Definitely thanks for the hard work on this though. Let's see what @BenjamenMeyer and @royfalk think.

**Comment by BenjamenMeyer** [comment]
Interesting...may be turn off OpenAL if there are errors? Not sure. we should certainly check the errors though.

**Comment by daviewales** [comment]
Yes, I didn't make a PR as I didn't expect this would be a good diff to apply without further research.

I did try disabling OpenAL by setting HAVE_AL to 0 in `engine/CMakeLists.txt`, but these functions still caused it to crash.
Is there another way to disable OpenAL?

**Comment by woodenshoe-wi** [comment]
I’m not that familiar with C++ but I see you have a “help wanted” label so here is my amateur attempt to fix this.

I don’t know if passing the ALCdevice to the checkAlcError(x) and checkAlcWarning(x) functions in the header file is written correctly, but that is the best I could figure out searching the internet for how to pass arguments in `#define` statements.

https://github.com/woodenshoe-wi/Vega-Strike-Engine-Source/commit/ac6f1d5a67c2945e6a94e09596cbbc88ed7ec554

I ran into this error in OpenALRenderer.cpp trying to run Vegastrike on Linux Mint, so it’s not specific to Window or macOS, but now I can’t reproduce the error either.

These changes won’t stop whatever is happening with OpenAL, but will log the problems at line 215 without throwing an exception.

**Comment by BenjamenMeyer** [comment]
seems like `checkAlError()` needs some work for when OpenAL is not enabled - or we need to make it a hard dependency

**Comment by woodenshoe-wi** [comment]
As I said before, I’m not a professional programmer, and I should add that I’m not that familiar with OpenAL either.

I started with what was already there, split the existing checkAlError function into two because from reading the OpenAL documentation it seems that there are two different functions for retrieving errors, alGetError and alcGetError, and also added the option of calling the functions in a way that would only log errors but not throw an exception.

All the files I modified were in a directory named OpenAL so I assumed they were only used if OpenAL was enabled.

As for what causes the actual error behind this issue, it could have something to do with OpenAL not being fully initialized before trying to use it, but I don’t quite understand what you mean by “hard dependency” because to me that sounds like a compile time thing, not a run time thing.

For what it’s worth the original error disappeared when running using gdb, so I do suspect something timing related.

I will gladly explain what I was trying to do, but I don’t know what should be done.

**Comment by BenjamenMeyer** [comment]
> As I said before, I’m not a professional programmer, and I should add that I’m not that familiar with OpenAL either.

That's the beauty of Open Source - its equally accessible to professional and non-professional alike, experienced and inexperienced.

> I started with what was already there, split the existing checkAlError function into two because from reading the OpenAL documentation it seems that there are two different functions for retrieving errors, alGetError and alcGetError, and also added the option of calling the functions in a way that would only log errors but not throw an exception.

Thanks for the insight. We may need to review the OpenAL functions we use to see which error method is more appropriate; its possible even that changes happened over time for one existing and now two, or one was more appropriate at one point and now the other is. So we probably have some maintenance there.

> All the files I modified were in a directory named OpenAL so I assumed they were only used if OpenAL was enabled.

@royfalk has a fairly good grasp of the code now; I'm getting there though a lot slower. That said, I don't think I would make that assumption yet. It's possibly true; but also very possibly not true.

> As for what causes the actual error behind this issue, it could have something to do with OpenAL not being fully initialized before trying to use it, but I don’t quite understand what you mean by “hard dependency” because to me that sounds like a compile time thing, not a run time thing.
> 
> For what it’s worth the original error disappeared when running using gdb, so I do suspect something timing related.

This is one reason I don't use debuggers because they always change the timing. It's better to rely on logs.

> I will gladly explain what I was trying to do, but I don’t know what should be done.

One way you could help is by reviewing the OpenAL functions we do call and seeing which OpenAL error method is appropriate when errors happen and submitting PRs for any needed changes. Sounds like you've done some investigatory work along these lines already. This would probably be a good first step towards helping find the root cause.

As with any open source project - help/PRs/documentation/bug reports are all welcomed.

**Comment by woodenshoe-wi** [comment]
I'll try to document my investigation so far...

I started with the `OpenAL error: unknown(Bad value)` message from the log file, but couldn't find much info searching the web except this issue on GitHub.

I found the `_checkAlErrorAt` function in OpenALHelpers.cpp that logs the error, but was puzzled by the error message.  The case statement that gives "Bad value" was ALC_INVALID_VALUE and I searched the web for more info and found the openal-soft implementation [alc.cpp](https://github.com/kcat/openal-soft/blob/master/alc/alc.cpp) which had several mentions of ALC_INVALID_VALUE but not in the `alcMakeContextCurrent` or `alcSuspendContext` functions.

On my computer, /usr/include/AL/alc.h defines ALC_INVALID_VALUE as 0xA004 and /usr/include/AL/al.h defines AL_INVALID_OPERATION as 0xA004

Since the `_checkAlErrorAt` function got the error from `alGetError()` I suspect it might be more appropriate to interpret the error as AL_INVALID_OPERATION

The openal-soft implementations of `alcMakeContextCurrent` and `alcSuspendContext` have `alcSetError(nullptr, ALC_INVALID_CONTEXT);` as the only error they return, so I'm wondering if a different thread is triggering an AL_INVALID_OPERATION error?

> One way you could help is by reviewing the OpenAL functions we do call and seeing which OpenAL error method is appropriate when errors happen

I had searched the source code for all calls to `checkAlError()` and changed the ones following the context management functions (alcOpenDevice, alcCloseDevice, alcCreateContext, alcMakeContextCurrent, alcProcessContext, alcSuspendContext and alcDestroyContext) in https://github.com/woodenshoe-wi/Vega-Strike-Engine-Source/commit/ac6f1d5a67c2945e6a94e09596cbbc88ed7ec554 but it's kind of a draft commit to discuss how the changes should be made.

It compiles and I was able to test that `checkAlError()` and `checkAlWarning()` work as intended by intentionally causing an OpenAL error, but I can't figure out how to intentionally cause an OpenAL context error to test the `checkAlcError(x)` and `checkAlcWarning(x)` functions.

I also don't know if going forward it would be a better strategy to log errors and let the program keep running if it can or throw an exception and stop when an error is encountered.

**Comment by stephengtuggy** [comment]
@daviewales @woodenshoe-wi @vincele @BenjamenMeyer Thanks for all your hard work on this and related issues. @woodenshoe-wi it looks like your code may well be on the right track. Just try passing the AL Context, instead of the AL Device, where the Context is called for. And to cause an OpenAL context error intentionally, perhaps you can try passing a really high context number (like 1 million or something). Other than that, it looks good.

**Comment by woodenshoe-wi** [comment]
Actually as I was thinking about it, I probably made things more complicated than necessary.

If you are interested in my changes, I would like to try making a simpler version.

Is there a particular branch / version that I should base the changes on?

**Comment by BenjamenMeyer** [comment]
> Actually as I was thinking about it, I probably made things more complicated than necessary.
> 
> If you are interested in my changes, I would like to try making a simpler version.

Yes. This seems to be related to some bugs that should be fixed.

> Is there a particular branch / version that I should base the changes on?

Please make this as a PR to our `master` branch which is where our active development happens. We split releases from here and if necessary backport changes to a current release.

**Comment by woodenshoe-wi** [comment]
I have the simplified commit ready.

I was going to compile it to make sure it worked…

And I discovered that Vega-Strike no longer supports Linux Mint Virginia…

I haven’t worked on this for over a year.

The new version is at woodenshoe-wi@48edd1a

What should the next step be?

**Comment by BenjamenMeyer** [comment]
@woodenshoe-wi yeah, we dropped Ubuntu Jammy support which is what Linux Mint Virginia is based on in #923 ... I can't remember the specific reason. @royfalk might have some insight there. I'm sure we had a good reason at the time.

Here's a few options:
1. Use a docker image to work with our master.
2. Upgrade to a newer version of Linux Mint based on 24.04 or in a couple months 26.04. (see note below)
3. Work on the 0.9.x branch (which should support 22.04 though we don't have CI support) and then we can help on migrating changes to ... then again just checking - 0.9.x we dropped it there in #1457 when syncing some OS detection logic that might have been unintentional (@stephengtuggy might be able to give some insight there); but you can see how to restore it with https://github.com/vegastrike/Vega-Strike-Engine-Source/pull/1457/commits/561c416c863fa135d54e5b7790e609787854bb24.

NOTE: I know option 2 is not ideal - far from it; just laying out options. I don't actually expect that to be the path forward that you take.

master may have required a dependency not available on 22.04 is my guess; unless there was good reason we should have kept CI support until we released 0.10.x since dependencies shouldn't change for an existing release but I can't recall the reason off hand.

If you do get it all working we'll evaluating whether to restore CI support on whichever branch you get working. Happy to work with you on that; though if it was for a dependency then support on master might be harder; but that shouldn't be an issue for 0.9.x.

**Comment by royfalk** [comment]
It was for boost JSON. Jammy has boost 1.74 and JSON was introduced in 1.75.
There's a lot of change happening now due to the introduction of imGUI and SDL3.
However, you can probably get away with just updating boost to 1.75 using apt magic and building manually.

**Comment by woodenshoe-wi** [comment]
> 2. Upgrade to a newer version of Linux Mint based on 24.04 or in a couple months 26.04. (see note below)

Actually… I have a computer that is currently running an even older version of Mint that I should probably upgrade anyway, so if a new version is coming out I think I’ll wait for the new version.

> If you do get it all working we'll evaluating whether to restore CI support on whichever branch you get working. Happy to work with you on that; though if it was for a dependency then support on master might be harder; but that shouldn't be an issue for 0.9.x.

I’m a self taught hobbyist programmer who was helping my brother install the game when we ran into this problem, (I don’t even know how to play the game myself), and the only reason I know about OpenAL is that I read the documentation years ago trying to get it working with Swift playgrounds on my iPad.

I’ll gladly take advantage of any learning opportunities I get, but I don’t really understand all the automation/AI stuff that GitHub is doing nowadays.  And although I’ve read books about C and Swift I don’t actually know C++

**Comment by BenjamenMeyer** [comment]
@woodenshoe-wi awesome.  Based on @royfalk 's comment -  I'd suggest figuring out how to install Boost 1.75 on your system and then building should hopefully just work. Feel free to join the [conversation on Gitter.im](https://matrix.to/#/#vegastrike_community:gitter.im) for more real-time-ish support.

**Comment by woodenshoe-wi** [comment]
I tried Boost 1.75 but ran into a problem with boost_python.  Boost 1.78 seems to work, but now cmake is complaining about not being able to find SDL2_IMAGE.

Two weeks ago Mint 22.3 Zena beta was released, when the stable version is released I do plan on upgrading one computer anyway.

**Comment by BenjamenMeyer** [comment]
> I tried Boost 1.75 but ran into a problem with boost_python. Boost 1.78 seems to work, but now cmake is complaining about not being able to find SDL2_IMAGE.

do you install `sdl2-image`?

**Comment by woodenshoe-wi** [comment]
I don't want to waste people's time trying to compile on an unsupported OS version, and it's not any extra work for me to upgrade a computer that should be upgraded anyway, but yes, for curiosity's sake I had tried to install sdl2-image.

```
sudo apt install libsdl2-image
...
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
E: Unable to locate package libsdl2-image

sudo apt install sdl2-image
...
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
E: Unable to locate package sdl2-image
```

**Comment by BenjamenMeyer** [comment]
@woodenshoe-wi I think the full package names will be:
```
libsdl2-image-2.0-0 - Image loading library for Simple DirectMedia Layer 2, libraries
libsdl2-image-dev - Image loading library for Simple DirectMedia Layer 2, development files
```

I think we do the `-dev` on in our CI. You can check the general linux build scripts for the Debian OSes we do support

**Comment by royfalk** [comment]
Yeah. Use `apt search sdl2 | grep -i image` to narrow the name down.

**Comment by woodenshoe-wi** [comment]
I guess that explains why I wasn't able to install `libsdl2-image` or `sdl2-image`, but not why I'm getting the cmake error.

```
CMake Error at libraries/imgui/CMakeLists.txt:32 (FIND_PACKAGE):
  By not providing "FindSDL2_IMAGE.cmake" in CMAKE_MODULE_PATH this project
  has asked CMake to find a package configuration file provided by
  "SDL2_IMAGE", but CMake did not find one.

  Could not find a package configuration file provided by "SDL2_IMAGE" with
  any of the following names:

    SDL2_IMAGEConfig.cmake
    sdl2_image-config.cmake

  Add the installation prefix of "SDL2_IMAGE" to CMAKE_PREFIX_PATH or set
  "SDL2_IMAGE_DIR" to a directory containing one of the above files.  If
  "SDL2_IMAGE" provides a separate development package or SDK, be sure it has
  been installed.
```
libsdl2-image-2.0-0 was already installed.
```
dpkg -s libsdl2-image-2.0-0
...
Package: libsdl2-image-2.0-0
Status: install ok installed
...
dpkg -s libsdl2-image-dev
...
Package: libsdl2-image-dev
Status: install ok installed
```

**Comment by stephengtuggy** [comment]
Huh. I don't know why you're getting that CMake error. That's weird.

**Comment by woodenshoe-wi** [comment]
So, I finally got around to installing Linux Mint 22.3 Zena and installed everything in the bootstrapOnLinuxMint section of bootstrap-on-linux.sh and when I tried to run cmake I got this...
```
CMake Error at CMakeLists.txt:465 (FIND_PACKAGE):
  By not providing "FindSDL3.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "SDL3", but
  CMake did not find one.

  Could not find a package configuration file provided by "SDL3" with any of
  the following names:

    SDL3Config.cmake
    sdl3-config.cmake

  Add the installation prefix of "SDL3" to CMAKE_PREFIX_PATH or set
  "SDL3_DIR" to a directory containing one of the above files.  If "SDL3"
  provides a separate development package or SDK, be sure it has been
  installed.
```
Now, in the bootstrap-on-linux.sh script it says "Linux Mint base Distros do NOT support SDL3" so this is expected, but do I have to disable something so that cmake will finish?

**Comment by stephengtuggy** [comment]
> So, I finally got around to installing Linux Mint 22.3 Zena and installed everything in the bootstrapOnLinuxMint section of bootstrap-on-linux.sh and when I tried to run cmake I got this...
> 
> ```
> CMake Error at CMakeLists.txt:465 (FIND_PACKAGE):
>   By not providing "FindSDL3.cmake" in CMAKE_MODULE_PATH this project has
>   asked CMake to find a package configuration file provided by "SDL3", but
>   CMake did not find one.
> 
>   Could not find a package configuration file provided by "SDL3" with any of
>   the following names:
> 
>     SDL3Config.cmake
>     sdl3-config.cmake
> 
>   Add the installation prefix of "SDL3" to CMAKE_PREFIX_PATH or set
>   "SDL3_DIR" to a directory containing one of the above files.  If "SDL3"
>   provides a separate development package or SDK, be sure it has been
>   installed.
> ```
> 
> Now, in the bootstrap-on-linux.sh script it says "Linux Mint base Distros do NOT support SDL3" so this is expected, but do I have to disable something so that cmake will finish?

@woodenshoe-wi you have two main options here:

1. Set `INSTALL_SDL3` and `INSTALL_SDL3_IMAGE` environment variables to truthy values such as "ON", "TRUE", or "1", so that CMake fetches SDL3 and SDL3_image using FetchContent_MakeAvailable instead of relying on the operating system package manager to provide them; or
2. Switch to the `0.10.x` branch, which does not require SDL3 or SDL3_image.

**Comment by stephengtuggy** [comment]
Given that you previously had issues with **SDL2**, my recommendation is option number 1. Hopefully that will get you going.

**Comment by woodenshoe-wi** [comment]
Finally got it compiled!
```
INSTALL_SDL3=1 INSTALL_SDL3_IMAGE=1 cmake ..
```
Worked and cmake completed with a warning,
```
Still need to generate dependency list
CMake Warning at engine/CMakeLists.txt:960 (MESSAGE):
  !! Unsupported Debian derivative
```
But compilation completed successfully on my second try.

On my first try I used `cmake --build ./build -j $(getconf _NPROCESSORS_ONLN)` as recommended in the README, but this computer has a 6 core processor (getconf _NPROCESSORS_ONLN returns 12) and only 8GB RAM.  Computer froze up, had to hold the power button to get it shut down, and when I tried to finish the build it gave errors so I deleted the build directory and started over using -j 4 this time.

Enough for tonight, will try some testing later.

**Comment by stephengtuggy** [comment]
@woodenshoe-wi Sweet!

I should update that part of the README. It's out of date compared to the build scripts, which now use a maximum of 8 cores for compilation for that very reason.

Anyway, I'm glad it compiled for you. Hopefully it works without major issues.

**Comment by woodenshoe-wi** [comment]
>... And to cause an OpenAL context error intentionally, perhaps you can try passing a really high context number (like 1 million or something). Other than that, it looks good.

I tried `checkAlcError((ALCdevice*)10000000);`(compiler complained when tried without the (ALCdevice*) cast) but no error messages in sight ☹️

https://github.com/woodenshoe-wi/Vega-Strike-Engine-Source/commit/48edd1a#diff-2caeae469afd891d0d5c19783ae7b60382aa29cc85f02641b2d7abe2220539dcR35

This is the line that I am worried about,
```
#define checkAlcError(x) Audio::__impl::OpenAL::_checkAlcErrorAt(x, __FILE__, __LINE__)
```
The other line that I'm deleting has (error) in parenthesis,
```
#define checkAlErrorCode(error) Audio::__impl::OpenAL::_checkAlErrorAt((error), __FILE__, __LINE__)
```
I don't know if I'm doing this correctly, and if this is wrong the result will be a *lack* of error messages.

**Comment by stephengtuggy** [comment]
Yeah, you probably shouldn’t delete those lines. They look important.

On Sat, Mar 7, 2026, at 12:48 PM, woodenshoe-wi wrote:
> *woodenshoe-wi* left a comment (vegastrike/Vega-Strike-Engine-Source#859) <https://github.com/vegastrike/Vega-Strike-Engine-Source/issues/859#issuecomment-4017315878>
>> ... And to cause an OpenAL context error intentionally, perhaps you can try passing a really high context number (like 1 million or something). Other than that, it looks good.
>> 
> I tried `checkAlcError((ALCdevice*)10000000);`(compiler complained when tried without the (ALCdevice*) cast) but no error messages in sight ☹️
> 
> ***@***.***#diff-2caeae469afd891d0d5c19783ae7b60382aa29cc85f02641b2d7abe2220539dcR35
> 
> This is the line that I am worried about,
> 
> `#define checkAlcError(x) Audio::__impl::OpenAL::_checkAlcErrorAt(x, __FILE__, __LINE__)
`
> The other line that I'm deleting has (error) in parenthesis,
> 
> `#define checkAlErrorCode(error) Audio::__impl::OpenAL::_checkAlErrorAt((error), __FILE__, __LINE__)
`

**Comment by woodenshoe-wi** [comment]
Searching the entire repo with grep only gives two results for 'checkAlErrorCode'
```
./engine/src/audio/renderers/OpenAL/OpenALRenderableSource.cpp:191:        checkAlErrorCode(error);
./engine/src/audio/renderers/OpenAL/OpenALHelpers.h:34:#define checkAlErrorCode(error) Audio::__impl::OpenAL::_checkAlErrorAt((error), __FILE__, __LINE__)
```
And I'm changing OpenALRenderableSource.cpp to use checkAlError(); instead here,
https://github.com/woodenshoe-wi/Vega-Strike-Engine-Source/commit/48edd1a#diff-8dcc8de3d9e70de5b7f2401bdeb7a684bfd8adab4b69a94d7b4345141d358c58L191

The only difference I could see between the checkAlErrorCode definition and the checkAlError definition is that checkAlErrorCode passes the error value from a previous `ALenum error = alGetError();` while checkAlError calls alGetError() in the definition.

The only reason that OpenALRenderableSource.cpp is checking for ALC_INVALID_ENUM is so that it can be ignored without crashing the program, and the comments indicate that it would be nice to log the event.

I'm removing the `throw Exception(error);` here, so no errors will cause an exception anymore.
https://github.com/woodenshoe-wi/Vega-Strike-Engine-Source/commit/48edd1a#diff-a323d7103b5c29b08922456401fe86887ae96522ef3eb4ec4ed7f1c7e740a365L84

If you still do want the program to throw an exception on errors I think it should only be enabled in debug mode.

I tried adding parenthesis around the x in the checkAlcError definition that I'm adding, but I still didn't get any error messages.  Do I have to look in a log file for these errors or enable logging?  It's been quite a while since I was working on this.

**Comment by stephengtuggy** [comment]
In software development, the number of times something occurs doesn't always directly correlate to its importance. Even if checkAlErrorCode is only called once, I think we should probably keep it.

Not crashing the program is a good goal. But we still need to handle AL errors somehow -- and not just in debug mode. How shall we do that? Maybe create one or more custom exception types, and throw and then catch those?

**Comment by stephengtuggy** [comment]
I definitely think the `x` is supposed to have parentheses around it in a macro like I assume you're adding. Triggering error messages may be difficult; I'm not immediately sure how to do that. But you can check the logs, which should be located under ~/.vegastrike/logs.

**Comment by woodenshoe-wi** [comment]
I added parentheses and changed x to a more descriptive alcdev in woodenshoe-wi@5340c08

I found the logs directory, but didn't find any audio related error messages there.  I can't remember how I was able to intentionally trigger an OpenAL error either.

The functions are supposed to always log the errors, I don't think throwing an exception is necessary for logging, but at this point I've forgotten how to test this.

As I was searching the web I came across this post https://github.com/kcat/openal-soft/issues/973#issuecomment-1940507788 saying that `alGetError` might have problems in multi-threaded applications and now I'm not sure that my changes will even fix the problem.

**Comment by stephengtuggy** [comment]
> I added parentheses and changed x to a more descriptive alcdev in [woodenshoe-wi/Vega-Strike-Engine-Source@5340c08](https://github.com/woodenshoe-wi/Vega-Strike-Engine-Source/commit/5340c08)

Looks good.
 
> I found the logs directory, but didn't find any audio related error messages there. I can't remember how I was able to intentionally trigger an OpenAL error either.

Okay. We'll have to try to figure out something.

> The functions are supposed to always log the errors, I don't think throwing an exception is necessary for logging, but at this point I've forgotten how to test this.

True. In terms of the logical control flow, though, the game might need to do something different in response when an error occurs.

> As I was searching the web I came across this post [kcat/openal-soft#973 (comment)](https://github.com/kcat/openal-soft/issues/973#issuecomment-1940507788) saying that `alGetError` might have problems in multi-threaded applications and now I'm not sure that my changes will even fix the problem.

While Vega Strike is multi-threaded, I am under the impression that it only uses a single thread for most OpenAL-related operations. I could be wrong though; we should verify this.

**Comment by woodenshoe-wi** [comment]
> While Vega Strike is multi-threaded, I am under the impression that it only uses a single thread for most OpenAL-related operations. I could be wrong though; we should verify this.

The original "residual error" that this issue is about seems to be coming from another thread.  I could understand getting "ALC_INVALID_CONTEXT" from `alcMakeContextCurrent` or `alcSuspendContext` but not "ALC_INVALID_VALUE" or "AL_INVALID_OPERATION" (both are defined as 0xA004, and could have been confused in the original checkAlError function)

> True. In terms of the logical control flow, though, the game might need to do something different in response when an error occurs.

As long as the game *doesn't* need to do anything differently when an error occurs, it shouldn't matter which thread's call to `checkAlError` logs the error, but it will be confusing for debugging if the source code file name and line number have nothing to do with what actually caused the error.

> Just try passing the AL Context, instead of the AL Device, where the Context is called for.

I didn't understand what this was about at first, but now I noticed that I do call `checkAlcError(alDevice);`
This does seem confusing, but the `checkAlcError` function passes the device to `alcGetError` and according to the documentation `alcGetError` needs a device and not a context.

From OpenAL Programmers Guide pdf
```
ALCenum alcGetError( ALCdevice *device );
```

alDevice is actually an ALCdevice
https://github.com/vegastrike/Vega-Strike-Engine-Source/blob/4cc213f634e7161f9a2f8052c69e172541f93d69/engine/src/audio/renderers/OpenAL/OpenALRenderer.cpp#L59

At this point I want to share any information that could be helpful that I can still remember, but if the code needs changes I think someone who is more comfortable writing in c++ and understands how throwing exceptions and error handling works needs to take over.

**Comment by stephengtuggy** [comment]
@woodenshoe-wi Fair enough. You've done a lot of legwork on this. Thanks for your contributions.
